orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/Stage.java

```diff
@@ -402,6 +403,106 @@ public Task taskById(String taskId) {
     }
   }
 
+  /**
+   * Computes all ancestor stages, including those of parent pipelines
+   *
+   * @return list of ancestor stages
+   */
+  public List<Stage> ancestorsWithParentPipelines() {
+    return Stage.getAncestors(this, this.execution);
+  }
+
+  private static List<Stage> getAncestors(Stage stage, Execution execution) {
+    List<Stage> allPriorStages = new ArrayList<>();
+
+    if (stage != null && !stage.getRequisiteStageRefIds().isEmpty()) {
+      List<Stage> previousStages =
+          execution.getStages().stream()
+              .filter(s -> stage.getRequisiteStageRefIds().contains(s.getRefId()))
+              .collect(toList());
+
+      Set<String> previousStageIds =
+          new HashSet<>(previousStages.stream().map(Stage::getId).collect(toList()));
+      List<Stage> syntheticStages =
+          execution.getStages().stream()
+              .filter(s -> previousStageIds.contains(s.getParentStageId()))
+              .collect(toList());
+
+      allPriorStages.addAll(previousStages);
+      allPriorStages.addAll(syntheticStages);
+
+      for (Stage s : previousStages) {
+        allPriorStages.addAll(getAncestors(s, execution));
+      }
+    } else if ((stage != null) && !Strings.isNullOrEmpty(stage.getParentStageId())) {
+      Optional<Stage> parent =
+          execution.getStages().stream()
+              .filter(s -> s.getId().equals(stage.getParentStageId()))
+              .findFirst();
+
+      if (!parent.isPresent()) {
+        throw new IllegalStateException(
+            "Couldn't find parent of stage "
+                + stage.getId()
+                + " with parent "
+                + stage.getParentStageId());
+      }
+
+      allPriorStages.add(parent.get());
+      allPriorStages.addAll(getAncestors(parent.get(), execution));
+    } else if ((execution.getType() == PIPELINE)
+        && (execution.getTrigger() instanceof PipelineTrigger)) {
+      PipelineTrigger parentTrigger = (PipelineTrigger) execution.getTrigger();
```

Comment:
This goes only one level back, correct? So your package must be created either in the existing pipeline or its parent, but not its grandparent.

Reply:
can we add a test case for this case if not present already : parent -> child -> grandchild

Reply:
I think this will go back multiple levels, getAncestors() recurses:

```
      if (parentPipelineStage.isPresent()) {
        allPriorStages.addAll(getAncestors(parentPipelineStage.get(), parentPipelineExecution));
      } else {
```

This could be problematic from a resource consumption standpoint, especially in cases I've seen within the community where users have massive chains of dependent pipelines.

I'd like to see parent pipeline traversal limited to fulfilling a search condition. So instead of "get all stages across all parents, then filter for $condition", "traverse ancestors until $condition".

Reply:
Yes, @asher is correct - this should recurse until root execution is found.

@asher : can you elaborate on this comment (we can chat in person as well - no rush). I agree it would be more efficient to only recurse until the condition is satisfied. But other than the efficiency are there any other reasons for it that I am not understanding?

Because the stages are returned in order (e.g. closest ancestor first) so the search downstream will find the closest match (assuming they traverse in order which currently all callers do).

To be clear: I like your suggestion and I think it will actually simplify some `findImage` code - just want to make sure I understand it completely since I am not super familiar with various pipeline designs you've seen from the community